test/files/neg/userdefined_apply.scala

```diff
@@ -0,0 +1,33 @@
+// NOTE: the companion inherits a public apply method from Function1!
+case class NeedsCompanion private (x: Int)
+
+object ClashNoSig { // ok
+  private def apply(x: Int) = if (x > 0) new ClashNoSig(x) else ???
+}
+case class ClashNoSig private (x: Int)
+
+// needs sig because overloaded -- TODO: actual error is double def
+object ClashOverloadNoSig {
+  private def apply(x: Int) = if (x > 0) new ClashOverloadNoSig(x) else apply("")
+  def apply(x: String): ClashOverloadNoSig = ???
+}
+case class ClashOverloadNoSig private (x: Int)
+
+// needs sig because it's recursive
```

Comment:
error message would ideally also be more clear in this case ("recursive method needs result type"), but the cure is the same.